gpii/node_modules/preferencesServer/test/cloudSafeCredTests.js

```diff
@@ -0,0 +1,300 @@
+    var data = JSON.parse(response);
+    jqUnit.assertEquals("There should be a prefsSafe", data.prefsSafe.type, "prefsSafe");
+    jqUnit.assertEquals("There should be a cloud cred", data.keys[0].type, "gpiiCloudSafeCredential");
+    jqUnit.assertEquals("There should be a gpii key", data.keys[1].type, "gpiiKey");
+};
+
+gpii.tests.preferencesServer.cloudSafeCred.put.buildSuccessTestDef = function (fixture) {
+    var expectedStatusCode = fixture.expectedStatusCode || 200;
+    return {
+        name: fixture.name,
+        expect: 8,
+        config: gpii.tests.preferencesServer.config,
+        components: {
+            putCloudCred: {
+                type: "kettle.test.request.http",
+                options: {
+                    path: "/add-cloud-credentials/" + fixture.url,
+                    method: "PUT",
+                    port: 8081,
+                    termMap: {
+                        prefsSafeId: fixture.prefsSafeId
+                    }
+                }
+            },
+            getFullSafe: {
+                type: "kettle.test.request.http",
+                options: {
+                    path: "/prefssafe-with-keys/" + fixture.url,
+                    method: "GET",
+                    port: 8081,
+                    termMap: {
+                        prefsSafeId: fixture.prefsSafeId
+                    }
+                }
+            },
+            unlockCloudSafe: {
+                type: "kettle.test.request.http",
+                options: {
+                    path: "/unlock-cloud-safe",
+                    method: "POST",
+                    port: 8081
+                }
+            }
+        },
+        sequence: [{
+            func: "{putCloudCred}.send",
+            args: fixture.putCloudCredBody
+        }, {
+            event: "{putCloudCred}.events.onComplete",
+            listener: "gpii.tests.preferencesServer.cloudSafeCred.testCloudSafeCreatePut",
+            args: ["{arguments}.0", fixture.putCloudCredBodyExpected, "{putCloudCred}.nativeResponse.statusCode", expectedStatusCode]
+        }, {
+            func: "{getFullSafe}.send"
+        }, {
+            event: "{getFullSafe}.events.onComplete",
+            listener: "gpii.tests.preferencesServer.cloudSafeCred.testCloudSafeCred",
+            args: ["{arguments}.0", fixture.getCloudCredFullSafeExpected, "{getFullSafe}.nativeResponse.statusCode", expectedStatusCode]
+        }, {
+            func: "{unlockCloudSafe}.send",
+            args: fixture.unlockCloudSafePost
+        }, {
+            event: "{unlockCloudSafe}.events.onComplete",
+            listener: "gpii.tests.preferencesServer.testResponse",
+            args: ["{arguments}.0", fixture.unlockCloudSafeExpected, "{unlockCloudSafe}.nativeResponse.statusCode", expectedStatusCode]
+        }]
+    };
+};
+
+gpii.tests.preferencesServer.cloudSafeCred.put.successFixtures = [
+    {
+        name: "PUT: Add a cloudsafe credential to an existing prefssafe",
```

Comment:
Would be nice to test failed cases such as when the given prefs safe id doesn't exist or the given user name or password is invalid.

Reply:
Added a test (and fixed the implementation!) for when the prefs safe does not exist. This is a test in the unlock section for the invalid username and password case. It uses the same semantics as successful tests with a json payload returned, so it's in there.